src/clr-angular/forms/combobox/combobox.ts

```diff
@@ -46,69 +72,228 @@ export function comboboxDomAdapterFactory(platformId: Object) {
+    public cdr: ChangeDetectorRef,
+    public commonStrings: ClrCommonStringsService,
     private toggleService: ClrPopoverToggleService,
-    private optionSelectionService: OptionSelectionService<T>,
-    private renderer: Renderer2,
-    private domAdapter: ComboboxDomAdapter
+    private positionService: ClrPopoverPositionService,
+    @Optional() private offsetService: ComboboxOffsetService,
+    private domAdapter: ComboboxDomAdapter,
+    private ariaService: AriaService,
+    private focusHandler: ComboboxFocusHandler
   ) {
-    // TODO: Might need to delay this to later in the lifecycle and use a BehaviorSubject instead
+    super(vcr, ClrComboboxContainer, injector, control, renderer, el);
+    if (control) {
+      control.valueAccessor = this;
+    }
+    // default to SingleSelectComboboxModel, in case the optional input [ClrMulti] isn't used
+    this.optionSelectionService.selectionModel = new SingleSelectComboboxModel<T>();
+
     this.initializeSubscriptions();
   }
 
-  private initializeSubscriptions(): void {
-    this.subscription = this.optionSelectionService.renderSelectionChanged.subscribe((option: ClrOption<T>) => {
-      this.renderSelection(option);
-    });
+  get openState(): boolean {
+    return this.toggleService.open;
   }
 
-  private renderSelection(selectedOption: ClrOption<T>): void {
-    if (this.input && selectedOption) {
-      this.domAdapter.clearChildren(this.input.nativeElement);
-      const clone: HTMLElement = this.domAdapter.cloneNode(selectedOption.elRef.nativeElement);
-      this.renderer.setAttribute(clone, 'contenteditable', 'false');
-      this.renderer.appendChild(this.input.nativeElement, clone);
+  public smartPosition: ClrPopoverPosition = {
+    axis: ClrAxis.VERTICAL,
+    side: ClrSide.AFTER,
+    anchor: ClrAlignment.START,
+    content: ClrAlignment.START,
+  };
+
+  loadingStateChange(state: ClrLoadingState): void {
+    this.optionSelectionService.loading = state === ClrLoadingState.LOADING;
+    this.positionService.realign();
+  }
+
+  private get disabled() {
+    return this.control && this.control.disabled;
+  }
+
+  unselect(item: T) {
+    if (!this.disabled) {
+      this.optionSelectionService.setSelected(item, false);
+      this.focusHandler.focus();
     }
   }
 
-  private registerPopoverIgnoredInput() {
-    if (this.input) {
-      // @TODO COMBOBOX: intentionally commented; resolve while merging the Combobox
-      //this.toggleService.registerIgnoredElement(this.input);
+  @Input('clrMulti')
+  set multiSelect(value: boolean) {
+    if (value) {
+      this.optionSelectionService.selectionModel = new MultiSelectComboboxModel<T>();
+    } else {
+      // in theory, setting this again should not cause errors even though we already set it in constructor,
+      // since the initial call to writeValue (caused by [ngModel] input) should happen after this
+      this.optionSelectionService.selectionModel = new SingleSelectComboboxModel<T>();
```

Comment:
If I read this correct, this line just ovewrites the selection model with a new object that has the same methods. Why not do nothing here?

Reply:
Well, if I read the code and comments right (this was inherited code to me):
- if we don't specify the `clrMulti` parameter - we set model to single in the constructor
- if there is a specified parameter - we overwrite it with either multi or single
- we overwrite the single above to be able to handle dynamic changes multi/single; which is highly unlikely to happen in practice, but we still do not want to fail if someone does it.